### PR TITLE
Use standardized followThis type for tuple leader/follower

### LINKED ISSUE #8313: Compiler error when zippering tuple and array in a forall loop
### Summary of Problem
A zippered `forall` loop over a tuple and array results in a (confusing) compiler error. Changing the leader (first argument to `zip()`) impacts the error message given.

### Steps to Reproduce

**Source Code:**

Promoted assignment works as expected
```chapel
var V = [0,0,0];
var T = (1,2,3);

V = T;
```

Serial zippered for loop works as expected
```chapel
var V = [0,0,0];
var T = (1,2,3);

for (v, t) in zip(V, T) do 
  v = t;
```

Parallel zippered `forall` loop with array as leader results in a compiler error
```chapel
var V = [0,0,0];
var T = (1,2,3);

// ChapelTuple.chpl:231: In iterator 'these':
// ChapelTuple.chpl:234: error: tuple indexing expression is not of integral type
forall (v, t) in zip(V, T) do
  v = t;
```

Parallel zippered `forall` loop with tuple as leader results in a different compiler error
```chapel
var V = [0,0,0];
var T = (1,2,3);

// error: unresolved access of 'range(int(64),bounded,false)' by '(1)'
forall (t, v) in zip(T, V) do
  v = t;
```

### Configuration Information

- Output of `chpl --version`: `chpl version 1.17.0 pre-release (05b5ccf)`

### PR DESCRIPTION
**Pull request by ben-albrecht**
Tuple leader/follower iterators now yield/expect tuples of 0-based ranges, instead of 1-based ranges. This conforms to the type of `followThis` used throughout the rest of `internal/modules`.

- Adds a test to lock in this behavior
- Adds a compile error to give a helpful error message when zipping a tuple with a multi-dimensional object
- Consolidates other tuple iteration tests to the same directory `types/tuples/these` to improve organization slightly

Closes #8313 

- [x] paratest: `[Summary: #Successes = 8238 | #Failures = 0]`

### CODE REVIEW

File: modules/internal/ChapelTuple.chpl [hunk 1]
```diff
@@ -209,29 +209,35 @@ module ChapelTuple {
...
 
   pragma "no doc"
-  iter _tuple.these(param tag:iterKind, followThis)
+  iter _tuple.these(param tag:iterKind, followThis: _tuple)
       where tag == iterKind.follower
   {
-    for i in followThis do yield this(i);
+    // Received followThis from a multi-dimensional object (e.g. a 2D array)
+    if followThis.size != 1 then
+      compilerError('Tuple cannot be zipped with a multi-dimensional object in a forall');
```
[the remark this answers is not in the document]
Comment (ben-albrecht) [reply modules/internal/ChapelTuple.chpl]: Will add error message tests. Let's try to narrow down a good error message - I'll ask for opinions from a broader audience.

Here are some ideas:

- "Tuple iteration is incompatible with leading iterator"
- "Iterators are Incompatible"
- "Tuple zipped with incompatible iterator" 
  - since an error would occur regardless of which iterator is leader

[the remark this answers is not in the document]
Comment (bradcray) [reply modules/internal/ChapelTuple.chpl]: I think the message that is here isn't so bad, and is attractive in that it's more specific and therefore easier to understand than the other proposals here.  To address @benharsh's concern, the iterator itself could potentially be tightened up to constrain it to only match against yielded tuples of ranges (I'm assuming that Ben's concern is that an associative domain of 2-tuples could/would have a leader that yielded 2-tuples and therefore would match against this call?  Note that the .translate() line is also likely to blow up in weird ways in that case...  I also suspect that many other follower iterators would have similar problems in cases like that...).

If we were to stick with the current assumption ("we're generating an error for cases that try to zip with multidimensional arrays"), I'd reword slightly to "tuples cannot be zipped with multidimensional objects."  If we were to go with something more general, I'd probably say "tuple zipped with incompatible iterator expression."

[the remark this answers is not in the document]
Comment (benharsh) [reply modules/internal/ChapelTuple.chpl]: My concern is mainly that ``followThis.size`` or ``followThis(1).type`` doesn't really say anything about what the leader is, at least in practice today. Until we have a better way of determining what the leader is, I'd prefer we have a more generic error message like your last suggestion.

I think it's up for debate whether a generic error message would be helpful for a user, but it could certainly be less confusing than mentioning multidimensional arrays where none may be involved.

[the remark this answers is not in the document]
Comment (ben-albrecht) [reply modules/internal/ChapelTuple.chpl]: > but it could certainly be less confusing than mentioning multidimensional arrays where none may be involved.

I agree. I simply hadn't considered irregular array types as other triggers to this compiler error. I would be fine with the error message: "tuple zipped with incompatible iterator expression."
